### ISSUE TITLE
Wheel installation fails when local path dependency introduced

### ISSUE THREAD

**Issue by stonelazy** [python-poetry/poetry#4868]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: MAC OS BigSur
- **Poetry version**: Poetry version 1.1.12
- **Reproducible code**:  https://github.com/stonelazy/poetry-path-issue 
- 
- ## Issue  
    - **.whl file generated using poetry build throws error when trying to install.** 
    - Steps to reproduce this issue 
        > git clone https://github.com/stonelazy/poetry-path-issue
        > cd poetry-path-issue
        > poetry build
        > pip install dist/appname-1-py3-none-any.whl -vvv    
     
    - Expected result
          - whl gets installed without any error.
    - Issue faced
          - Exception trace thrown when invoking pip install of the .whl file   
    - Issue Description
           - This issue occurs only when a another local path dependency is kept. Suppose the line [L13](https://github.com/stonelazy/poetry-path-issue/blob/main/pyproject.toml#L13) is removed, then there is no issue.
           - Including of this local path is correct according to https://python-poetry.org/docs/dependency-specification/#path-dependencies
```     
ERROR: Exception:
Traceback (most recent call last):
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3021, in _dep_map
    return self.__dep_map
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2815, in __getattr__
    raise AttributeError(attr)
AttributeError: _DistInfoDistribution__dep_map

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3101, in __init__
    super(Requirement, self).__init__(requirement_string)
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/packaging/requirements.py", line 117, in __init__
    raise InvalidRequirement(f"Invalid URL: {req.url}")
pip._vendor.packaging.requirements.InvalidRequirement: Invalid URL: appname-pkg3

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_internal/cli/base_command.py", line 173, in _main
    status = self.run(options, args)
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_internal/cli/req_command.py", line 203, in wrapper
    return func(self, options, args)
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_internal/commands/install.py", line 315, in run
    requirement_set = resolver.resolve(
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/resolver.py", line 94, in resolve
    result = self._result = resolver.resolve(
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py", line 472, in resolve
    state = resolution.resolve(requirements, max_rounds=max_rounds)
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py", line 366, in resolve
    failure_causes = self._attempt_to_pin_criterion(name)
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py", line 212, in _attempt_to_pin_criterion
    criteria = self._get_updated_criteria(candidate)
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/resolvelib/resolvers.py", line 202, in _get_updated_criteria
    for requirement in self._p.get_dependencies(candidate=candidate):
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/provider.py", line 197, in get_dependencies
    return [r for r in candidate.iter_dependencies(with_requires) if r is not None]
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/provider.py", line 197, in <listcomp>
    return [r for r in candidate.iter_dependencies(with_requires) if r is not None]
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_internal/resolution/resolvelib/candidates.py", line 250, in iter_dependencies
    requires = self.dist.requires() if with_requires else ()
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/pkg_resources/__init__.py", line 2736, in requires
    dm = self._dep_map
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3023, in _dep_map
    self.__dep_map = self._compute_dependencies()
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3033, in _compute_dependencies
    reqs.extend(parse_requirements(req))
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3094, in parse_requirements
    yield Requirement(line)
  File "/Users/sudharsan-2598/opt/anaconda3/envs/npsp6/lib/python3.9/site-packages/pip/_vendor/pkg_resources/__init__.py", line 3103, in __init__
    raise RequirementParseError(str(e))
pip._vendor.pkg_resources.RequirementParseError: Invalid URL: appname-pkg3
```
   
- Tree of the repo   
```
├── README.md
├── appname
│   └── pkg1
│       ├── __init__.py
│       └── sample.py
├── appname-pkg3
│   ├── appname
│   │   └── pkg3
│   │       ├── __init__.py
│   │       └── sample.py
│   └── pyproject.toml
├── dist
│   ├── appname-1-py3-none-any.whl
│   └── appname-1.tar.gz
├── error.txt
├── poetry.lock
└── pyproject.toml
```
   
- Expected Result:
    - Installation gets completed and that i would be able to do following the python import.
    ```   
    from appname.pkg3 import sample
    from appname.pkg1 import sample
    ```   

- Why should this be a poetry issue and not something wrong with the code ?   
    - If i do poetry lock and then poetry install then i don't face any issue.

**Comment by michaelcredera** [comment]
I get this behavior as well - building a wheel with a local path dependency results in a wheel that can't be installed on other systems (even when wheels for that path dependency have been previously built and installed).

**Comment by KholdStare** [comment]
Currently setting up a monorepo with python services/libraries, and need to work around these current limitations.

At the end of the day, poetry supports local paths for dependencies so I think there needs to be a sane outcome when running poetry build in this case. IMHO, it should most likely package all the local dependencies up into the built .whl file - that's the simplest solution. If something is local, and we're building a module to be installed somewhere, it makes sense to bundle any non-external dependencies with the build, because there is no other way to get them.

I mentioned the same in https://github.com/python-poetry/poetry/issues/1168

**Comment by abn** [comment]
@stonelazy I suspect your issue is related to https://github.com/pypa/pip/issues/6658#issuecomment-562973835. For `poetry-core`'s part, atleast for now, we do write the dependency specification as required by the wheel `METADATA`. Outside of that it is the PEP 517 front end that needs to handle it.

Also even in a case where the above is fixed, your code will need the following snippet in order for the wheel to include the package as a vendored dependency.

```toml
include = [
    { path = "appname-pkg3", format = ["sdist", "wheel"] }
]
```

This will ensure the source is vendored into your wheel when built.

@michaelcredera for your issue, the relative paths are problematic. When wheels are being built under PEP-517, the source is expected to be moved into an isolated environment at build time. This means your relative paths will need to hold even when your project source is moved to an ephemeral build directory. Unlike the OP's case yours is trickier. The general question of "what should this be relative to?" is still an open question within the packaging group. See [here](https://discuss.python.org/t/what-is-the-correct-interpretation-of-path-based-pep-508-uri-reference/2815) for more on this.

The alternative approach for this case is to build `../../libs/analysis` into a wheel and use `pip install --find-links /wheelhouse project.whl` on install. This might not be applicable in all cases.

**Comment by KholdStare** [comment]
> When wheels are being built under PEP-517, the source is expected to be moved into an isolated environment at build time. This means your relative paths will need to hold even when your project source is moved to an ephemeral build directory. Unlike the OP's case yours is trickier. The general question of "what should this be relative to?" is still an open question within the packaging group. See [here](https://discuss.python.org/t/what-is-the-correct-interpretation-of-path-based-pep-508-uri-reference/2815) for more on this.

At the risk of sounding ignorant, I'm speaking from a user/"peasant" perspective of someone trying to make things work with the available tools. It just doesn't seem like leaving it up to pip to resolve build the dependencies - it would be a wrong level of abstraction no? When `poetry build` finishes I would think the user expects a self contained package that knows how to get its dependencies. Only include one "level" (just the current package), and then leaving the rest of the job up to pip seems very strange. Imagine if there's 3 levels of dependencies - the first level is the wheel, the second level would be up to pip to make an on-the-fly build using poetry, and then again recursively try and build the 3 level. I am not sure I understand why pip has to be in the middle of this at that point if the references are relative/local.

IMHO, it should be a special case, or at least there should be a commandline switch or configuration to do this with poetry all during the build step itself, and not unnecessarily postpone half of the process to a separate tool (i.e. pip). It just overcomplicates something that users intuitively expect.

**Comment by TBBle** [comment]
I wouldn't want my pathed dependencies to be vendored in, personally. I'd rather pathed dependencies be turned into package-dependencies, perhaps with a version-equality exactly matching the version-at-time-of-build of the pathed dependencies. My approach would be to `poetry build` each of my packages in the monorepo, and then make all the resulting wheels available so that installing the top-level one (or ones) pulls in the appropriate lower-level wheel from the same directory.

So similar to what https://github.com/python-poetry/poetry/issues/4868#issuecomment-1040485726 expected, that pre-installing the resulting depended-upon wheel would work correctly.

**Comment by KholdStare** [comment]
Understood - I can agree with not vendoring in by default. However,

>  I'd rather pathed dependencies be turned into package-dependencies, perhaps with a version-equality exactly matching the version-at-time-of-build of the pathed dependencies.

This is quite cumbersome and non-intuitive from a usability perspective. It means having a pathed dependency requires a host of conventions and scripts that come with it to make it work. I.e. if I have `dependency= {path = "../dependency", develop = true}` listed in my poetry project, then I need:
- to make sure I also have a versioned dependency `dependency = "0.1.1"`
- I need to keep that version manually synced up with the version in the other directory (more scripts or manual work here)
- I cannot use the result of `poetry build` directly
- I need external scripts outside of poetry that will "stitch up the build" in CI.

Otherwise IMO the path dependency is "incomplete" as a feature because it has all these caveats attached to it - it's not useable. At the very least these should be expressed in the documentation to make sure users know what to expect. Hopefully it would be possible to do this without any of the extra steps.

**Comment by TBBle** [comment]
I was actually thinking that `poetry build` would be doing the first two dot points, because when it looks at the pathed dependency, it can see the version of the pointed-to package _generate_ that `==` dependency as part of the resulting setup.py pretty trivially.

I agree that it's not a workflow for every occasion: It works well for "I run `poetry build on every package in parallel in my CI system, upload them to a directory, and then `COPY` that directory into a DockerFile where I call `pip install` for the particular top-level packages being installed in this case", but vendoring as you describe is a better workflow for `poetry build -o X.whl && pip install X.whl` for a particular top-level package X.

The flow I want to use also works for "I build each package and push it to our PyPI repo in parallel", particularly if there are consumers of the libraries (not the root packages) who're not in the same monorepo, and hence need a normally exported package _anyway_. This flow means that both the monorepo output and the external users are using the same wheels.

That said, if vendoring into a single wheel is implemented and reliable, I'd probably use it in preference to the current approach, which is exposing all the source into the containers and `poetry install` there. It'd be a significant improvement, even if it's not _my_ perfection vision.

**Comment by michaelcredera** [comment]
> The alternative approach for this case is to build `../../libs/analysis` into a wheel and use `pip install --find-links /wheelhouse project.whl` on install. This might not be applicable in all cases.

@abn Pip seems to still have trouble with relative paths - for me this resulted in the same "Invalid URL" error:

```
pip._vendor.pkg_resources.RequirementParseError: Invalid URL: ../../libs/analysis
```

Process:

1. Build `analysis` package which has only pypi dependencies
2. Build `other` package which depends on `../../libs/analysis` 
3. Install `analysis` wheel - no issues
4. Install `other` wheel - breaks due to invalid requirement for `other` - the relative path "url"

**Comment by abn** [comment]
The root cause of the issue has been addressed in https://github.com/python-poetry/poetry/issues/4868#issuecomment-1061303442.

As there is no standard mechanism in wheels to handle relative path dependencies, and as any resolution in pip and/or wheel spec should address the issue, I am closing this.

Regarding the broader use cases for monorepos/subprojects, see #936 and #2270.

If you would like help regarding your project and/or guidence around workarounds, I suggest a [discussion](https://github.com/python-poetry/poetry/discussions) be started.

**Comment by Jack1082013** [comment]
So keep having issues with the wheel dependency for pysha3


Building wheel for pysha3 (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [18 lines of output]
      running bdist_wheel                               running build
      running build_py
      creating build
      creating build/lib.linux-aarch64-cpython-311
      copying sha3.py -> build/lib.linux-aarch64-cpython-311                                              running build_ext
      building '_pysha3' extension
      creating build/temp.linux-aarch64-cpython-311
      creating build/temp.linux-aarch64-cpython-311/Modules
      creating build/temp.linux-aarch64-cpython-311/Modules/_sha3
      aarch64-linux-android-clang -DNDEBUG -g -fwrapv -O3 -Wall -fstack-protector-strong -O3 -fstack-protector-strong -O3 -fPIC -DPY_WITH_KECCAK=1 -I/data/data/com.termux/files/usr/include/python3.11 -c Modules/_sha3/sha3module.c -o build/temp.linux-aarch64-cpython-311/Modules/_sha3/sha3module.o
      In file included from Modules/_sha3/sha3module.c:20:                                                Modules/_sha3/backport.inc:78:10: fatal error: 'pystrhex.h' file not found
         78 | #include "pystrhex.h"
            |          ^~~~~~~~~~~~
      1 error generated.                                error: command '/data/data/com.termux/files/usr/bin/aarch64-linux-android-clang' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for pysha3
  Running setup.py clean for pysha3
Failed to build pysha3
ERROR: Could not build wheels for pysha3, which is required to install pyproject.toml-based projects
Retrying installation of merkletools...
Collecting merkletools
  Using cached merkletools-1.0.3-py3-none-any.whl
Collecting pysha3>=1.0b1 (from merkletools)
  Using cached pysha3-1.0.2.tar.gz (829 kB)
  Preparing metadata (setup.py) ... done
Building wheels for collected packages: pysha3
  Building wheel for pysha3 (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [18 lines of output]
      running bdist_wheel
      running build
      running build_py
      creating build
      creating build/lib.linux-aarch64-cpython-311
      copying sha3.py -> build/lib.linux-aarch64-cpython-311
      running build_ext
      building '_pysha3' extension
      creating build/temp.linux-aarch64-cpython-311
      creating build/temp.linux-aarch64-cpython-311/Modules
      creating build/temp.linux-aarch64-cpython-311/Modules/_sha3
      aarch64-linux-android-clang -DNDEBUG -g -fwrapv -O3 -Wall -fstack-protector-strong -O3 -fstack-protector-strong -O3 -fPIC -DPY_WITH_KECCAK=1 -I/data/data/com.termux/files/usr/include/python3.11 -c Modules/_sha3/sha3module.c -o build/temp.linux-aarch64-cpython-311/Modules/_sha3/sha3module.o
      In file included from Modules/_sha3/sha3module.c:20:
      Modules/_sha3/backport.inc:78:10: fatal error: 'pystrhex.h' file not found
         78 | #include "pystrhex.h"
            |          ^~~~~~~~~~~~
      1 error generated.
      error: command '/data/data/com.termux/files/usr/bin/aarch64-linux-android-clang' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.                ERROR: Failed building wheel for pysha3
  Running setup.py clean for pysha3
Failed to build pysha3
ERROR: Could not build wheels for pysha3, which is required to install pyproject.toml-based projects
Failed to install merkletools.
dpkg: error processing package zeronet (--configure):
 installed zeronet package post-installation script subprocess returned error exit status 1
Setting up wuzz (0.5.0-4) ...
Errors were encountered while processing:
 zeronet
E: Sub-process /data/data/com.termux/files/usr/bin/dpkg returned an error code (1)

I have used $pip install " " --no-cache-dir.  it works sometimes. But I would like to solve the issue. The old python was made obsolete and the new has no id

**Comment by TBBle** [comment]
@Jack1082013 Your problem does not appear to have any connection to this issue, or in fact any connection with Poetry at all.

I guess that you've hit https://github.com/Tierion/pymerkletools/pull/20 and should use the approach described there, and follow up with zeronet (the thing you appear to be trying to install, that is pulling in pymerketools, that is pulling in pysha3) if you still can't get it to work.

**Comment by Jack1082013** [comment]
The issue has started coming up when I try to install anything on my
terminal it's not just with zero net but I will look into your suggested
fix and get back to you with a reply
